File: packages/client/src/app/controllers/game-play-controller/game-play.controller.ts
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { ActionData } from '@app/classes/actions/action-data';
import GameUpdateData from '@app/classes/communication/game-update-data';
import { Message } from '@app/classes/communication/message';
import { HTTP_ABORT_ERROR } from '@app/constants/controllers-errors';
import SocketService from '@app/services/socket-service/socket.service';
import { TilePlacement } from '@common/models/tile-placement';
import { BehaviorSubject, Observable, Subject } from 'rxjs';
import { environment } from 'src/environments/environment';

@Injectable({
    providedIn: 'root',
})
export class GamePlayController {
    private gameUpdate$ = new BehaviorSubject<GameUpdateData>({});
    private newMessage$ = new BehaviorSubject<Message | null>(null);
    private tilePlacement$ = new BehaviorSubject<TilePlacement[]>([]);
    private actionDone$ = new Subject<void>();

    constructor(private http: HttpClient, private readonly socketService: SocketService) {
        this.configureSocket();
    }

    sendAction(gameId: string, action: ActionData): void {
        const endpoint = `${environment.serverUrl}/games/${gameId}/players/action`;
        this.http.post(endpoint, action).subscribe(() => {
            this.actionDone$.next();
        });
    }
    replaceVirtualPlayerByObserver(gameId: string, virtualPlayerNumber: number): void {
        const endpoint = `${environment.serverUrl}/games/${gameId}/players/replace`;
        this.http.post(endpoint, { virtualPlayerNumber }).subscribe();
    }
    sendMessage(gameId: string, message: Message): void {
        const endpoint = `${environment.serverUrl}/games/${gameId}/players/message`;
        this.http.post(endpoint, message).subscribe();
    }

    sendError(gameId: string, message: Message): void {
        const endpoint = `${environment.serverUrl}/games/${gameId}/players/error`;
        this.http.post(endpoint, message).subscribe();
    }

    handleReconnection(gameId: string, newPlayerId: string): void {
        const endpoint = `${environment.serverUrl}/games/${gameId}/players/reconnect`;
        this.http.post(endpoint, { newPlayerId }).subscribe();
    }

    handleDisconnection(gameId: string): void {
        const endpoint = `${environment.serverUrl}/games/${gameId}/players/disconnect`;
        // When reloading the page, a disconnect http request is fired on destruction of the game-page component.
        // In the initialization of the game-page component, a reconnect request is made which does not allow the
        // server to send a response, triggering a Abort 0  error code which is why we catch it if it this this code
        this.http.delete(endpoint, { observe: 'response' }).subscribe(this.handleDisconnectResponse, this.handleDisconnectError);
    }

    handleTilePlacement(gameId: string, tilePlacement: TilePlacement[]): void {
        const endpoint = `${environment.serverUrl}/games/${gameId}/squares/place`;
        this.http.post(endpoint, { tilePlacement }).subscribe();
    }

    observeGameUpdate(): Observable<GameUpdateData> {
        return this.gameUpdate$.asObservable();
    }

    observeNewMessage(): Observable<Message | null> {
        return this.newMessage$.asObservable();
    }

    observeActionDone(): Observable<void> {
        return this.actionDone$.asObservable();
    }

    observeTilePlacement(): Observable<TilePlacement[]> {
        return this.tilePlacement$.asObservable();
    }

    private configureSocket(): void {
        this.socketService.on('gameUpdate', (newData: GameUpdateData) => {
            this.gameUpdate$.next(newData);
        });
        this.socketService.on('newMessage', (newMessage: Message) => {
            this.newMessage$.next(newMessage);
        });
        this.socketService.on('tilePlacement', (tilePlacement: TilePlacement[]) => {
            this.tilePlacement$.next(tilePlacement);
        });
    }

    private handleDisconnectResponse(): void {
        return;
    }

    private handleDisconnectError(error: { message: string; status: number }): void {
        if (error.status !== HTTP_ABORT_ERROR) throw new Error(error.message);
    }
}

File: packages/mobile/lib/controllers/game-play.controller.dart
import 'dart:convert';

import 'package:http_interceptor/http/http.dart';
import 'package:mobile/classes/actions/action-data.dart';
import 'package:mobile/classes/game/game-message.dart';
import 'package:mobile/classes/game/game-update.dart';
import 'package:mobile/classes/http/ResponseResult.dart';
import 'package:mobile/constants/endpoint.constants.dart';
import 'package:mobile/constants/game.constants.dart';
import 'package:mobile/constants/socket-events/game-events.dart';
import 'package:mobile/locator.dart';
import 'package:mobile/services/socket.service.dart';
import 'package:rxdart/rxdart.dart';

import '../services/client.dart';

class GamePlayController {
  GamePlayController._privateConstructor() {
    configureSocket();
  }

  static final GamePlayController _instance =
      GamePlayController._privateConstructor();

  factory GamePlayController() {
    return _instance;
  }

  SocketService socketService = getIt.get<SocketService>();

  final String baseEndpoint = GAME_ENDPOINT;
  PersonnalHttpClient httpClient = getIt.get<PersonnalHttpClient>();
  InterceptedHttp get http => httpClient.http;

  String? currentGameId;

  Stream<ResponseResult> get actionDoneEvent => _actionDone$.stream;

  Stream<GameUpdateData> get gameUpdateEvent => gameUpdate$.stream;

  Stream<GameMessage?> get messageEvent => gameMessage$.stream;

  final BehaviorSubject<GameUpdateData> gameUpdate$ =
      BehaviorSubject<GameUpdateData>();
  final BehaviorSubject<GameMessage?> gameMessage$ =
      BehaviorSubject<GameMessage?>.seeded(null);
  final PublishSubject<ResponseResult> _actionDone$ =
      PublishSubject<ResponseResult>();

  Future<void> sendAction(ActionData actionData) async {
    Uri endpoint = Uri.parse("$baseEndpoint/$currentGameId/players/action");
    http.post(endpoint, body: jsonEncode(actionData));
  }

  Future<void> leaveGame() async {
    Uri endpoint = Uri.parse("$baseEndpoint/$currentGameId/players/leave");
    await http.delete(endpoint).then((_) {
      currentGameId = null;
    });
  }

  Future<void> replaceVirtualPlayer(int playerNumber) async {
    Uri endpoint = Uri.parse("$baseEndpoint/$currentGameId/players/replace");
    var data = {'virtualPlayerNumber': playerNumber};
    await http.post(endpoint, body: jsonEncode(data));
  }

  void configureSocket() {
    socketService.on(GAME_UPDATE_EVENT_NAME, (dynamic newData) {
      gameUpdate$.add(GameUpdateData.fromJson(newData));
    });
    socketService.on(GAME_MESSAGE_EVENT_NAME, (dynamic gameMessage) {
      GameMessage message = GameMessage.fromJson(gameMessage);
      gameMessage$.add(message);

      _processActionDone(message);
    });
  }

  void _processActionDone(GameMessage message) {
    if (message.senderId == SYSTEM_ID) {
      _actionDone$.add(ResponseResult.success());
    } else if (message.senderId == SYSTEM_ERROR_ID) {
      _actionDone$.add(ResponseResult.error());
    }
  }
}
